File: 2022/1/go/main.go
// https://adventofcode.com/2022/day/1
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	part1()
	part2()
}

func part1() {
	contentBytes, err := os.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}
	content := string(contentBytes)

	elvesStrs := strings.Split(content, "\r\n\r\n")
	elves := make([][]int, len(elvesStrs))
	for i, elfStr := range elvesStrs {
		itemStrs := strings.Split(elfStr, "\r\n")
		elves[i] = make([]int, len(itemStrs))
		for j, itemStr := range itemStrs {
			itemCalories, err := strconv.Atoi(itemStr)
			if err != nil {
				panic(err)
			}
			elves[i][j] = itemCalories
		}
	}

	elfSums := make([]int, len(elves))
	for i, elf := range elves {
		for _, itemCalories := range elf {
			elfSums[i] += itemCalories
		}
	}
	sort.Ints(elfSums)
	result := elfSums[len(elfSums)-1]
	if result != 69836 {
		panic(result)
	}
	fmt.Println(result)
}

func part2() {
	contentBytes, err := os.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}
	content := string(contentBytes)

	elvesStrs := strings.Split(content, "\r\n\r\n")
	elves := make([][]int, len(elvesStrs))
	for i, elfStr := range elvesStrs {
		itemStrs := strings.Split(elfStr, "\r\n")
		elves[i] = make([]int, len(itemStrs))
		for j, itemStr := range itemStrs {
			itemCalories, err := strconv.Atoi(itemStr)
			if err != nil {
				panic(err)
			}
			elves[i][j] = itemCalories
		}
	}

	elfSums := make([]int, len(elves))
	for i, elf := range elves {
		for _, itemCalories := range elf {
			elfSums[i] += itemCalories
		}
	}
	sort.Ints(elfSums)
	top3 := elfSums[len(elfSums)-3:]
	result := top3[0] + top3[1] + top3[2]
	if result != 207968 {
		panic(result)
	}
	fmt.Println(result)
}

File: 2022/1/C#/C#/main.cs
// https://adventofcode.com/2022/day/1

main();

static void main()
{
    part1();
    part2();
}

static void part1()
{
    //string content = File.ReadAllText("../../../test-input.txt");
    //int? expectedResult = 24000;

    string content = File.ReadAllText("../../../input.txt");
    int? expectedResult = 69836;

    List<int> allCalories = new() { 0 };
    foreach (string line in content.Split("\r\n"))
    {
        if (line.Length > 0)
            allCalories[^1] += int.Parse(line);
        else
            allCalories.Add(0);
    }

    allCalories.Sort();
    int mostCalories = allCalories.Last();

    Console.WriteLine($"Part 1 result: {mostCalories}");
    if (expectedResult != null && expectedResult != mostCalories)
    {
        throw new Exception($"Incorrect result: {mostCalories}");
    }
}

static void part2()
{
    //string content = File.ReadAllText("../../../test-input.txt");
    //int? expectedResult = 45000;

    string content = File.ReadAllText("../../../input.txt");
    int? expectedResult = 207968;

    List<int> allCalories = new() { 0 };
    foreach (string line in content.Split("\r\n"))
    {
        if (line.Length > 0)
            allCalories[^1] += int.Parse(line);
        else
            allCalories.Add(0);
    }

    allCalories.Sort();
    int top3Calories = allCalories.GetRange(allCalories.Count() - 3, 3).Sum();

    Console.WriteLine($"Part 2 result: {top3Calories}");
    if (expectedResult != null && expectedResult != top3Calories)
    {
        throw new Exception($"Incorrect result: {top3Calories}");
    }
}
